### PR TITLE
Feature/62 seed products table with some sample products

### LINKED ISSUE #62: Seed Products table with some sample products
Add a migration to build some products in the Products table so we can have data when testing

### PR DESCRIPTION
**Pull request by IliaIliev94**
Closes #62

* Added a ModelBuilderExtensions class containing a SeedMethod responsible for seeding the database with sample product entries.
* Created a seperate method called SeedProducts returning an array of Products to pass to the SeedMethod (this is done so that if in the future we decide to seed another table the SeedMethod won't get too long).

### CODE REVIEW

**Comment by valincius** [comment]
This looks like a really clunky way to seed some data, especially just sample data for testing. I usually go the route of adding a blank migration then writing the inserts & deletes myself for seeding data like this - we can wait for someone elses' opinion but this is the way I've always done it and seems a lot simpler to me:

```c#
public partial class SeedProducts : Migration
    {
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.Sql(@"
                INSERT INTO Products (Id, Name, Description, Price, ImageUrl)
                VALUES
                ('4e434a20-7661-458e-9be0-994b01272f85', 'ScrewDriver', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 5.99, 'https://via.placeholder.com/150'),
                ('511fa7ab-f70b-45a1-827b-d9214fc80fd3', 'AMD Ryzen 7 5900X', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 499.99, 'https://via.placeholder.com/150'),
                ('66ec0969-5566-4752-83e6-828e047a2c0e', 'A Rock', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 0.99, 'https://via.placeholder.com/150'),
                ('7f2090cc-c768-4ee8-8595-65a8d6555678', '5-Piece Gardening Kit', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 34.99, 'https://via.placeholder.com/150'),
                ('9ca58986-b5f5-4106-bf76-cf084f8ffb7f', 'Box of Chocolates', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 9.99, 'https://via.placeholder.com/150')
            ");
        }

        protected override void Down(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.Sql(@"
                DELETE FROM Products
                WHERE Id IN(
                    '4e434a20-7661-458e-9be0-994b01272f85',
                    '511fa7ab-f70b-45a1-827b-d9214fc80fd3',
                    '66ec0969-5566-4752-83e6-828e047a2c0e',
                    '7f2090cc-c768-4ee8-8595-65a8d6555678',
                    '9ca58986-b5f5-4106-bf76-cf084f8ffb7f'
                )
            ");
        }
    }
```

**Comment by chrisK00** [comment]
> This looks like a really clunky way to seed some data, especially just sample data for testing. I usually go the route of adding a blank migration then writing the inserts & deletes myself for seeding data like this - we can wait for someone elses' opinion but this is the way I've always done it and seems a lot simpler to me:
> 
> ```cs
> public partial class SeedProducts : Migration
>     {
>         protected override void Up(MigrationBuilder migrationBuilder)
>         {
>             migrationBuilder.Sql(@"
>                 INSERT INTO Products (Id, Name, Description, Price, ImageUrl)
>                 VALUES
>                 ('4e434a20-7661-458e-9be0-994b01272f85', 'ScrewDriver', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 5.99, 'https://via.placeholder.com/150'),
>                 ('511fa7ab-f70b-45a1-827b-d9214fc80fd3', 'AMD Ryzen 7 5900X', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 499.99, 'https://via.placeholder.com/150'),
>                 ('66ec0969-5566-4752-83e6-828e047a2c0e', 'A Rock', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 0.99, 'https://via.placeholder.com/150'),
>                 ('7f2090cc-c768-4ee8-8595-65a8d6555678', '5-Piece Gardening Kit', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 34.99, 'https://via.placeholder.com/150'),
>                 ('9ca58986-b5f5-4106-bf76-cf084f8ffb7f', 'Box of Chocolates', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 9.99, 'https://via.placeholder.com/150')
>             ");
>         }
> 
>         protected override void Down(MigrationBuilder migrationBuilder)
>         {
>             migrationBuilder.Sql(@"
>                 DELETE FROM Products
>                 WHERE Id IN(
>                     '4e434a20-7661-458e-9be0-994b01272f85',
>                     '511fa7ab-f70b-45a1-827b-d9214fc80fd3',
>                     '66ec0969-5566-4752-83e6-828e047a2c0e',
>                     '7f2090cc-c768-4ee8-8595-65a8d6555678',
>                     '9ca58986-b5f5-4106-bf76-cf084f8ffb7f'
>                 )
>             ");
>         }
>     }
> ```

I usually just have a static Seed class with a static method that i use on startup to insert items if the db doesnt have anything. I combine that with using a realistic data generator like Bogus or json generator.

**Comment by IliaIliev94** [comment]
> 
> 
> > This looks like a really clunky way to seed some data, especially just sample data for testing. I usually go the route of adding a blank migration then writing the inserts & deletes myself for seeding data like this - we can wait for someone elses' opinion but this is the way I've always done it and seems a lot simpler to me:
> > ```cs
> > public partial class SeedProducts : Migration
> >     {
> >         protected override void Up(MigrationBuilder migrationBuilder)
> >         {
> >             migrationBuilder.Sql(@"
> >                 INSERT INTO Products (Id, Name, Description, Price, ImageUrl)
> >                 VALUES
> >                 ('4e434a20-7661-458e-9be0-994b01272f85', 'ScrewDriver', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 5.99, 'https://via.placeholder.com/150'),
> >                 ('511fa7ab-f70b-45a1-827b-d9214fc80fd3', 'AMD Ryzen 7 5900X', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 499.99, 'https://via.placeholder.com/150'),
> >                 ('66ec0969-5566-4752-83e6-828e047a2c0e', 'A Rock', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 0.99, 'https://via.placeholder.com/150'),
> >                 ('7f2090cc-c768-4ee8-8595-65a8d6555678', '5-Piece Gardening Kit', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 34.99, 'https://via.placeholder.com/150'),
> >                 ('9ca58986-b5f5-4106-bf76-cf084f8ffb7f', 'Box of Chocolates', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 9.99, 'https://via.placeholder.com/150')
> >             ");
> >         }
> > 
> >         protected override void Down(MigrationBuilder migrationBuilder)
> >         {
> >             migrationBuilder.Sql(@"
> >                 DELETE FROM Products
> >                 WHERE Id IN(
> >                     '4e434a20-7661-458e-9be0-994b01272f85',
> >                     '511fa7ab-f70b-45a1-827b-d9214fc80fd3',
> >                     '66ec0969-5566-4752-83e6-828e047a2c0e',
> >                     '7f2090cc-c768-4ee8-8595-65a8d6555678',
> >                     '9ca58986-b5f5-4106-bf76-cf084f8ffb7f'
> >                 )
> >             ");
> >         }
> >     }
> > ```
> 
> I usually just have a static Seed class with a static method that i use on startup to insert items if the db doesnt have anything. I combine that with using a realistic data generator like Bogus or json generator.

Something like [this](https://www.learnentityframeworkcore.com/migrations/seeding#before-entity-framework-core-2.1)?

**Comment by valincius** [comment]
> > > This looks like a really clunky way to seed some data, especially just sample data for testing. I usually go the route of adding a blank migration then writing the inserts & deletes myself for seeding data like this - we can wait for someone elses' opinion but this is the way I've always done it and seems a lot simpler to me:
> > > ```cs
> > > public partial class SeedProducts : Migration
> > >     {
> > >         protected override void Up(MigrationBuilder migrationBuilder)
> > >         {
> > >             migrationBuilder.Sql(@"
> > >                 INSERT INTO Products (Id, Name, Description, Price, ImageUrl)
> > >                 VALUES
> > >                 ('4e434a20-7661-458e-9be0-994b01272f85', 'ScrewDriver', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 5.99, 'https://via.placeholder.com/150'),
> > >                 ('511fa7ab-f70b-45a1-827b-d9214fc80fd3', 'AMD Ryzen 7 5900X', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 499.99, 'https://via.placeholder.com/150'),
> > >                 ('66ec0969-5566-4752-83e6-828e047a2c0e', 'A Rock', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 0.99, 'https://via.placeholder.com/150'),
> > >                 ('7f2090cc-c768-4ee8-8595-65a8d6555678', '5-Piece Gardening Kit', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 34.99, 'https://via.placeholder.com/150'),
> > >                 ('9ca58986-b5f5-4106-bf76-cf084f8ffb7f', 'Box of Chocolates', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla vestibulum lacus augue, a tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus.', 9.99, 'https://via.placeholder.com/150')
> > >             ");
> > >         }
> > > 
> > >         protected override void Down(MigrationBuilder migrationBuilder)
> > >         {
> > >             migrationBuilder.Sql(@"
> > >                 DELETE FROM Products
> > >                 WHERE Id IN(
> > >                     '4e434a20-7661-458e-9be0-994b01272f85',
> > >                     '511fa7ab-f70b-45a1-827b-d9214fc80fd3',
> > >                     '66ec0969-5566-4752-83e6-828e047a2c0e',
> > >                     '7f2090cc-c768-4ee8-8595-65a8d6555678',
> > >                     '9ca58986-b5f5-4106-bf76-cf084f8ffb7f'
> > >                 )
> > >             ");
> > >         }
> > >     }
> > > ```
> > 
> > 
> > I usually just have a static Seed class with a static method that i use on startup to insert items if the db doesnt have anything. I combine that with using a realistic data generator like Bogus or json generator.
> 
> Something like [this](https://www.learnentityframeworkcore.com/migrations/seeding#before-entity-framework-core-2.1)?

Eh no, that's an old way to do it. The way you're doing it just fine actually.

File: src/Congo.WebApi/Extensions/ModelBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                        "tempor enim sollicitudin eu. Pellentesque odio nulla, maximus id mattis posuere, semper ac massa. Donec quis " +
+                        "lectus eros. Sed ac fringilla nunc. Proin lacus nisi, fermentum quis interdum nec, blandit in sapien. Phasellus " +
+                        "bibendum volutpat tortor, a sagittis sapien finibus id. Etiam urna est, sagittis eget orci et, finibus luctus lacus. ",
+                        Price = (decimal)NextDouble(random, 0.99, 999.99),
+                        ImageUrl = "https://via.placeholder.com/150",
+                    };
+                }
+            }
+        }
+        private static double NextDouble(Random random, double min, double max) => random.NextDouble() * (max - min) + min;
```
Comment (valincius) [inline src/Congo.WebApi/Extensions/ModelBuilderExtensions.cs]: Can you move this method someplace else? Maybe make a MathExtension class and add it as an extension method on `Random`.

**Review by valincius — APPROVED** [review]

**Pull request merged**